File: go/bst.go
package btree

import (
	"fmt"
	"strings"
)

// Node represents a node in the binary search tree
type BSTNode[T comparable] struct {
	Value T
	Left  *BSTNode[T]
	Right *BSTNode[T]
}

// BST represents the binary search tree
type BST[T comparable] struct {
	root *BSTNode[T]
	less func(a, b T) bool // Custom comparator
}

// NewBST creates a new binary search tree with a custom comparator
func NewBST[T comparable](less func(a, b T) bool) *BST[T] {
	return &BST[T]{
		root: nil,
		less: less,
	}
}

// Insert adds a new value to the BST
func (t *BST[T]) Insert(value T) {
	t.root = t.insert(t.root, value)
}

func (t *BST[T]) insert(node *BSTNode[T], value T) *BSTNode[T] {
	if node == nil {
		return &BSTNode[T]{Value: value}
	}

	if t.less(value, node.Value) {
		node.Left = t.insert(node.Left, value)
	} else if t.less(node.Value, value) {
		node.Right = t.insert(node.Right, value)
	}

	return node
}

// Find searches for a value in the BST
func (t *BST[T]) Find(value T) bool {
	return t.find(t.root, value)
}

func (t *BST[T]) find(node *BSTNode[T], value T) bool {
	if node == nil {
		return false
	}

	if value == node.Value {
		return true
	}

	if t.less(value, node.Value) {
		return t.find(node.Left, value)
	}
	return t.find(node.Right, value)
}

// Delete removes a value from the BST
func (t *BST[T]) Delete(value T) {
	t.root = t.delete(t.root, value)
}

func (t *BST[T]) delete(node *BSTNode[T], value T) *BSTNode[T] {
	if node == nil {
		return nil
	}

	if t.less(value, node.Value) {
		node.Left = t.delete(node.Left, value)
	} else if t.less(node.Value, value) {
		node.Right = t.delete(node.Right, value)
	} else {
		// Node to delete found

		// Case 1: Leaf node
		if node.Left == nil && node.Right == nil {
			return nil
		}

		// Case 2: Node with only one child
		if node.Left == nil {
			return node.Right
		}
		if node.Right == nil {
			return node.Left
		}

		// Case 3: Node with two children
		// Find the smallest value in the right subtree (successor)
		minNode := t.findMinNode(node.Right)
		node.Value = minNode.Value
		// Delete the successor
		node.Right = t.delete(node.Right, minNode.Value)
	}

	return node
}

// findMinNode finds the node with minimum value in the subtree
func (t *BST[T]) findMinNode(node *BSTNode[T]) *BSTNode[T] {
	current := node
	for current.Left != nil {
		current = current.Left
	}
	return current
}

// InorderTraversal returns the inorder traversal of the BST
func (t *BST[T]) InorderTraversal() []T {
	var result []T
	t.inorder(t.root, &result)
	return result
}

func (t *BST[T]) inorder(node *BSTNode[T], result *[]T) {
	if node != nil {
		t.inorder(node.Left, result)
		*result = append(*result, node.Value)
		t.inorder(node.Right, result)
	}
}

// String returns a string representation of the BST
func (t *BST[T]) String() string {
	var sb strings.Builder
	t.printTree(t.root, 0, &sb)
	return sb.String()
}

func (t *BST[T]) printTree(node *BSTNode[T], level int, sb *strings.Builder) {
	if node != nil {
		t.printTree(node.Left, level+1, sb)
		//sb.WriteString(strings.Repeat("    ", level))
		sb.WriteString(fmt.Sprintf("%v ", node.Value))
		t.printTree(node.Right, level+1, sb)
	}
}

File: cpp/bst.cpp
// To keep the code cleaner, break the declaration from the impl.

template<typename NodeValType>
shared_ptr<Node<NodeValType>> BST<NodeValType>::Insert(const NodeValType& nodeValue) noexcept {

	if (!root) {
		root = std::make_shared <Node<NodeValType>>(nodeValue);
		return root;
	}

	return InsertNode_Internal(root, nodeValue);
}

template<typename NodeValType>
bool BST<NodeValType>::Delete(const NodeValType& nodeValue) noexcept {
	const auto& [updatedNode, didDelete]{ DeleteNode_Internal(root, nodeValue) };

	// if we deleted the root update the root node.
	if (!root) {
		root = updatedNode;
	}

	return didDelete;
}

template<typename NodeValType>
shared_ptr<Node<NodeValType>> BST<NodeValType>::Find(const NodeValType& findNodeValue) noexcept {
	return FindNode_Internal(root, findNodeValue);
}

template<typename NodeValType>
void BST<NodeValType>::OutputTree() noexcept {
	OutputTree_Internal(root);
	cout << endl;
}

template<typename NodeValType>
string BST<NodeValType>::OutputTreeString() noexcept {

	string result;
	OutputTreeString_Internal(root, result);
	return result;
}

template<typename NodeValType>
std::shared_ptr<Node<NodeValType>> BST<NodeValType>::InsertNode_Internal(const std::shared_ptr<Node<NodeValType>>& node, const NodeValType& nodeValue) noexcept {

	if (!node) {
		return make_shared<Node<NodeValType>>(nodeValue);
	}

	// Recurse the tree to insert
	if (nodeValue < node->value) {
		node->left = InsertNode_Internal(node->left, nodeValue);
	}
	else if (nodeValue > node->value) {
		node->right = InsertNode_Internal(node->right, nodeValue);
	}

	return node;
}

template<typename NodeValType>
pair<shared_ptr <Node<NodeValType>>, bool> BST<NodeValType>::DeleteNode_Internal(const std::shared_ptr<Node<NodeValType>>& node, const NodeValType& nodeValue) noexcept {

	if (!node) {
		return { node, false };
	}

	// recurse to find the node
	if (nodeValue < node->value) {
		const auto& [updatedNode, result] = DeleteNode_Internal(node->left, nodeValue);
		node->left = updatedNode;
		if (!result) return { node, result };
	}
	else if(nodeValue > node->value) {
		const auto& [updatedNode, result] = DeleteNode_Internal(node->right, nodeValue);
		node->right = updatedNode;
		if (!result) return { node, result };
	}
	else {
		// node value found

		// either one child or no children
		// we do not delete the node but instead let it go out of scope
		// to be destructed
		if (!node->left) {
			return { node->right, true };
		}
		else if (!node->right) {
			return { node->left, true };
		}

		// a node with two children, find the inorder successor. The
		// smallest value in the right subtree
		auto temp = minValueNode(node->right);
		node->value = temp->value;
		const auto& [updatedNode, result] = DeleteNode_Internal(node->right, temp->value);
		node->right = updatedNode;
	}

	return { node, true };
}

template<typename NodeValType>
std::shared_ptr <Node<NodeValType>> BST<NodeValType>::minValueNode(const std::shared_ptr <Node<NodeValType>> node) noexcept {
	auto current = node;
	while (current && current->left) {
		current = current->left;
	}

	return current;
}

template<typename NodeValType>
shared_ptr<Node<NodeValType>> BST<NodeValType>::FindNode_Internal(const std::shared_ptr<Node<NodeValType>>& node, const NodeValType& findNodeValue) noexcept {

	if (!node || node->value == findNodeValue)
		return node;
	
	if (findNodeValue < node->value) {
		return FindNode_Internal(node->left, findNodeValue);
	}

	return FindNode_Internal(node->right, findNodeValue);
}

template<typename NodeValType>
void BST<NodeValType>::OutputTree_Internal(const shared_ptr<Node<NodeValType>>& node) noexcept {

	// output using in order traversal.
	if (node) {
		OutputTree_Internal(node->left);
		cout << node->value << " ";
		OutputTree_Internal(node->right);
	}
}

template<typename NodeValType>
void BST<NodeValType>::OutputTreeString_Internal(const shared_ptr<Node<NodeValType>>& node, string& output) noexcept {
	
	if (node) {
		OutputTreeString_Internal(node->left, output);
		output += to_string(node->value) + " ";
		OutputTreeString_Internal(node->right, output);
	}
}